### PR TITLE
Fix: Stretchy text issue when nested on flex containers.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Fixes an issue where Stretch text does not works well if nested inside flex based layouts like a stack whose width is limited for example by a column.

To fix the issue, we verify if the parent is a flex element, if yes, we use the parent width to constraint fit text (still taking into account padding) as for flex layouts the the descent could in theory grow indefinitely. For all other cases the previous logic is the same. 

cc: @henriqueiamarino

### Before
Screenshot 2025-11-28 at 19 31 05

### After
Screenshot 2025-11-28 at 19 30 53

## Testing Instructions
Open the editor past the following markup in the code editor:
```
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:group {"layout":{"type":"flex","orientation":"vertical"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>col 1</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"fitText":true} -->
<p class="has-fit-text">Stretch</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>col 2</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```
Verify the "Stretch" behavior is the expected one.

### CODE REVIEW

**Comment by ellatrix** [comment]
@jorgefilipecosta Since you added the backport label, I'll add it the board for consideration.

**Comment by jorgefilipecosta** [comment]
> @jorgefilipecosta Since you added the backport label, I'll add it the board for consideration.

Thank you ☺️

**Comment by t-hamano** [comment]
The Gutenberg package will not be synced in the RC4 release, so let's consider shipping it in a point release.

**Comment by t-hamano** [comment]
Perhaps the trickiest part is when the parent element has padding, which will cause the text to overflow:

image

We could probably solve this by checking the padding of both the text element and its parent element, but text elements can be further nested, so we may need to recursively check the parent's padding 🤔

**Comment by oandregal** [comment]
I've tested current `trunk` and I don't seem to reproduce the issue? Or perhaps I haven't fully understood it. These are my testing steps:

- copy the text in the issue description
- type something in the stretchy paragraph

My expectation was that it'd overflow the container, but it didn't (which is good).

https://github.com/user-attachments/assets/73d179ac-94d2-4f5e-9bfe-892105bd29a5

**Comment by t-hamano** [comment]
@oandregal Your environment seems a little out of date. Although "Stretchy Paragraph" is displayed in the block sidebar, the stretch variation has been removed in #73890.

In my environment, I can reproduce the problem as follows:

https://github.com/user-attachments/assets/4e7c7b0e-1b51-4276-9f70-5bdce0c5dad8

**Comment by jorgefilipecosta** [comment]
I rebased the PR and I retested that the fix is still needed.
When pasting the example code on trunk things don't work, while on this branch it works.

**Comment by t-hamano** [comment]
Unfortunately, if the parent or ancestor block has left or right padding, the text will overflow. See https://github.com/WordPress/gutenberg/pull/73652#issuecomment-3737340852

This is the AI-suggested code, but it's an approach that **recursively traverses the parents and accurately calculates the available width**. This is probably quite a redundant logic and could be improved.

<details><summary>Details</summary>

```javascript
function getAvailableWidth( element ) {
	const elementStyle = window.getComputedStyle( element );
	const elementPaddingLeft = parseFloat( elementStyle.paddingLeft ) || 0;
	const elementPaddingRight = parseFloat( elementStyle.paddingRight ) || 0;

	// Start with the element's own client width
	let availableWidth = element.clientWidth;

	// If the element is inside a flex container, we need to consider
	// the parent's constraints more carefully
	let parent = element.parentElement;

	while ( parent ) {
		const parentStyle = window.getComputedStyle( parent );
		const display = parentStyle.display;
		const isFlexContainer = display === 'flex' || display === 'inline-flex';

		if ( isFlexContainer ) {
			// For flex containers, calculate available width based on
			// the parent's client width minus padding
			const parentPaddingLeft =
				parseFloat( parentStyle.paddingLeft ) || 0;
			const parentPaddingRight =
				parseFloat( parentStyle.paddingRight ) || 0;
			const parentClientWidth = parent.clientWidth;

			// Calculate how much space is available from the parent
			const parentAvailableWidth =
				parentClientWidth - parentPaddingLeft - parentPaddingRight;

			// If we're in a flex container, the available width might be
			// constrained by the parent's available space
			// Use the minimum of current available width and parent's available width
			availableWidth = Math.min( availableWidth, parentAvailableWidth );

			// Also consider if there are siblings that might affect the layout
			// For flex containers, we need to be more conservative
			const flexDirection = parentStyle.flexDirection || 'row';
			if ( flexDirection === 'row' || flexDirection === 'row-reverse' ) {
				// In row direction, we need to consider the parent's width
				availableWidth = Math.min(
					availableWidth,
					parentAvailableWidth
				);
			}
		} else {
			// For non-flex containers, check if parent has padding that affects us
			const parentPaddingLeft =
				parseFloat( parentStyle.paddingLeft ) || 0;
			const parentPaddingRight =
				parseFloat( parentStyle.paddingRight ) || 0;

			if ( parentPaddingLeft > 0 || parentPaddingRight > 0 ) {
				// If parent has padding, we need to consider it
				const parentClientWidth = parent.clientWidth;
				const parentAvailableWidth =
					parentClientWidth - parentPaddingLeft - parentPaddingRight;

				// Use the minimum of current available width and parent's available width
				availableWidth = Math.min(
					availableWidth,
					parentAvailableWidth
				);
			}
		}

		// Move up the DOM tree
		parent = parent.parentElement;

		// Stop if we've reached the body or html element
		if (
			! parent ||
			parent === document.body ||
			parent === document.documentElement
		) {
			break;
		}
	}

	// Subtract the element's own padding
	return availableWidth - elementPaddingLeft - elementPaddingRight;
}

function findOptimalFontSize( textElement, applyFontSize ) {
	// ...

	// Calculate the actual available width considering flex containers and padding
	const availableWidth = getAvailableWidth( textElement );

	while ( minSize <= maxSize ) {
		// ...

		// Check if text fits within the available width
		const fitsWidth = textWidth <= availableWidth;

		// ...
	}
	range.detach();

	return bestSize;
}
```
</details> 

cc @mcsf if you have any ideas!

**Comment by mcsf** [comment]
> cc @mcsf if you have any ideas!

I appreciate that you thought of me for this sort of stuff, but I don't think there's any room left in my mind. 😅

**Comment by t-hamano** [comment]
This PR was featured in today's 6.9.1 bug scrub.

> Unfortunately, if the parent or ancestor block has left or right padding, the text will overflow. See https://github.com/WordPress/gutenberg/pull/73652#issuecomment-3737340852

Ideally, we should address this issue, but for now, I'm starting to think it might be okay to move forward with this PR.

**Comment by jorgefilipecosta** [comment]
> This PR was featured in today's 6.9.1 bug scrub.
> 
> > Unfortunately, if the parent or ancestor block has left or right padding, the text will overflow. See [#73652 (comment)](https://github.com/WordPress/gutenberg/pull/73652#issuecomment-3737340852)
> 
> Ideally, we should address this issue, but for now, I'm starting to think it might be okay to move forward with this PR.

I'm having a look into the padding issue. Seeing If there is some simple fix.

**Comment by jorgefilipecosta** [comment]
Hi @t-hamano I think the padding issue on flex is fixed, and this should be ready.

**Comment by t-hamano** [comment]
@jorgefilipecosta Thanks for the update!

Unfortunately, checking the current block and the parent block isn't sufficient. The flex container can be deeply nested or ancestor blocks may have padding.

See the screenshot below.

image

We'll probably need to loop all the way up to the document root to find the truly available width.

**Comment by jorgefilipecosta** [comment]
> @jorgefilipecosta Thanks for the update!
> 
> Unfortunately, checking the current block and the parent block isn't sufficient. The flex container can be deeply nested or ancestor blocks may have padding.
> 
> See the screenshot below.
> 
> image
> We'll probably need to loop all the way up to the document root to find the truly available width.


Good catch, 
I guess it happens when the ancestors are a chain of flex containers so we probably can loop while the ancestor is flex and sum the paddings. I will make some tries.

**Comment by t-hamano** [comment]
There is an unsteady code generated by AI 😄: https://github.com/WordPress/gutenberg/pull/73652#issuecomment-3754879690

**Comment by jorgefilipecosta** [comment]
> There is an unsteady code generated by AI 😄: [#73652 (comment)](https://github.com/WordPress/gutenberg/pull/73652#issuecomment-3754879690)

Oh sorry missed that suggestion, I will look into it 👍

File: packages/block-editor/src/utils/fit-text-utils.js [hunk 1]
```diff
@@ -36,21 +51,23 @@ function findOptimalFontSize( textElement, applyFontSize ) {
 		// Check if text fits within the element's width and is not
 		// overflowing into the padding area.
 		const fitsWidth =
-			textElement.scrollWidth <= textElement.clientWidth &&
-			textWidth <= textElement.clientWidth - paddingLeft - paddingRight;
+			textElement.scrollWidth <= referenceElement.clientWidth &&
+			// textWidth <= textElement.parentElement.clientWidth &&
+			textWidth <=
+				referenceElement.clientWidth - paddingLeft - paddingRight;
```
Comment (t-hamano) [inline packages/block-editor/src/utils/fit-text-utils.js:56]:
```suggestion
			textElement.scrollWidth <= referenceElement.clientWidth &&
			textWidth <=
				referenceElement.clientWidth - paddingLeft - paddingRight;
```

Nitpick:

**Review by t-hamano — APPROVED** [review]
LGTM! The text now fits correctly even when nested deeply. Let's backport this to 6.9.1.

https://github.com/user-attachments/assets/8c1a4b88-4a89-4655-bf9a-16795bf354ca

**Comment by jorgefilipecosta** [comment]
Thank you a lot for the reviews @t-hamano!

**Comment by t-hamano** [comment]
This PR has been backported to the `wp/6.9` branch. #74806
